gfs/benchmark/performance_test.go

```diff
@@ -0,0 +1,190 @@
+	Architecture: ARCHITECTURE,
+}
+
+func TestMain(m *testing.M) {
+	// TO:DO How to handle errors on the set up
+	if err := os.MkdirAll("../temp_dfs_storage", os.ModePerm); err != nil {
+		log.Fatal(err)
+	}
+
+	// Start up Master Server
+	go master.InitMasterServer(masterServerPort, NUM_CHUNK_SERVERS, chunkServerPortBase)
+
+	// Start up Chunkservers
+	for i := 0; i < NUM_CHUNK_SERVERS; i++ {
+		go chunkserver.InitBenchmarkingChunkServer(chunkServerPortBase+i, CS_BENCHMARK_CONFIG)
+	}
+
+	log.Println("Creating benchmark output files")
+	err := os.MkdirAll(CLIENT_BENCHMARK_CONFIG.OutputDir, os.ModePerm)
+	if err != nil {
+		log.Printf("err: %s", err)
+	}
+
+	time.Sleep(2 * time.Second) //Arbitrary Number
+
+	exitVal := m.Run()
+
+	// _ = os.RemoveAll("../temp_dfs_storage")
+
+	os.Exit(exitVal)
+}
+
+func Test_Single3000ByteCreateWriteReadRemove(t *testing.T) {
+
+	for _, file := range INPUT_FILES {
+
+		inputFile := file
+		for i := 0; i < NTRIALS; i++ {
+
+			c1, err := client.NewBenchmarkingClient(masterServerPort, CLIENT_BENCHMARK_CONFIG)
+			defer c1.MasterConn.Close()
+			assert.NoError(t, err)
+
+			targetFileName := "files/" + inputFile
+
+			data, err := ioutil.ReadFile(targetFileName)
+			if err != nil {
+				log.Printf("error reading file %s: %s", targetFileName, err)
+			}
+
+			filename := inputFile
+			c1.Create(filename)
+			c1.Write(filename, 0, data)
+
+			readBuffer := make([]byte, len(data))
+
+			c1.Read(filename, 0, readBuffer)
+
+			c1.Remove(filename)
+
+			match := bytes.Compare(data, readBuffer)
+			assert.True(t, match == 0, "ERROR: input and outputs do not match. Invalidate benchmark results.")
+		}
+	}
+
+}
+
+// NOTE: this does not run successfully at the moment. possibly bugs...?
+
+// func Test_Multi3000ByteCreateWriteReadRemove(t *testing.T) {
+
+// 	var F = []string{"textfile_500b.txt", "textfile_64kb.txt", "textfile_5mb.txt"}
+
+// 	for _, file := range F {
+
+// 		inputFile := file
+// 		for i := 0; i < NTRIALS; i++ {
+
+// 			var client_wg sync.WaitGroup
+// 			for clientNum := 0; clientNum < 2; clientNum++ {
+// 				client_wg.Add(1)
+// 				go func() {
+
+// 					c1, err := client.NewBenchmarkingClient(masterServerPort, CLIENT_BENCHMARK_CONFIG)
+// 					defer c1.MasterConn.Close()
+// 					assert.NoError(t, err)
+
+// 					targetFileName := "files/" + inputFile
+
+// 					data, err := ioutil.ReadFile(targetFileName)
+// 					if err != nil {
+// 						log.Printf("error reading file %s: %s", targetFileName, err)
+// 					}
+
+// 					filename := inputFile
+// 					c1.Create(filename)
+// 					c1.Write(filename, 0, data)
+
+// 					readBuffer := make([]byte, len(data))
+
+// 					c1.Read(filename, 0, readBuffer)
+
+// 					// c1.Remove(filename)
+
+// 					match := bytes.Compare(data, readBuffer)
+// 					assert.True(t, match == 0, "ERROR: input and outputs do not match. Invalidate benchmark results.")
+// 					client_wg.Done()
+
+// 				}()
+
+// 			}
+// 			client_wg.Wait()
+
+// 		}
+// 	}
+
+// }
```

Comment:
I tried doing a multi-client performace test and we get nil pointer reference. Could be caused by a race